### PR TITLE
[Experimental] ONNI v2

### PR DESCRIPTION
**Pull request by a127a127**
Ready for review:
- Refactor Interpreter to an InterpreterPass.
- Add Two passes for onni:
    1. CountOperatorPass: Operator counts statistics
    2. OnnxOptPass: Run onnx optimizer
- Add more verbose level for debugging or benchmarking
  1. level 1:
    Print memory usage and overall time.
  2. level 2:
    Print the final ONNC IR before inference.
  3. level 3:
    Print the operator counts.
    Print the per layer inference time.
  4. level 4:
    Print internal memory layout.
- Add two flags:
  1. --dry-run: Do not do the inference, just print statistics.   
  2. --onnx-opt: Enable onnx optimizer
- Do not allocate input memory for interpreter.

### CODE REVIEW

File: tools/onni/InterpreterPass.cpp [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  Interpreter interpreter(runtime_context);
+
+  // XXX: Use onnc-runtime to handle memory
+  char *heap = NULL;
+
+  // XXX: Use Pass or something to get internal memory size
+  uint64_t max_size = 0;
+  for (ComputeOperand *co : pModule.getComputeOperands()) {
+    if (ComputeMemOperand *mem = dyn_cast<ComputeMemOperand>(co)) {
+      Value *v = co->getValue();
```
Comment (cycheng) [inline tools/onni/InterpreterPass.cpp]: how about mem->getValue();

File: tools/onni/InterpreterPass.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      if (mem->isInput()) {
+        // XXX: Multiple inputs
```
Comment (cycheng) [inline tools/onni/InterpreterPass.cpp]: Use assert? assert(inputNum == 1 && "TODO: Support multiple inputs);

File: tools/onni/InterpreterPass.cpp [hunk 3]
```diff
@@ -94,6 +103,60 @@ Pass::ReturnType InterpreterPass::runOnModule(Module &pModule)
...
+  }
+
+  if (m_Verbose >= 4) {
+    std::ios old_state(nullptr);
+    old_state.copyfmt(outs());
+
+    // TODO: Refactor this. We need a table printer.
+    size_t val_len = 8;
+    for (ComputeOperator &cm : *pModule.getRootComputeGraph()) {
+      if (dyn_cast<InputOperator>(&cm)) continue;
```
Comment (cycheng) [inline tools/onni/InterpreterPass.cpp]: How about add a function:
bool skipInMemStatistics(ComputeOperator *pCm) {
  if (isa<InputOperator>(pCm) ||
      isa<...> || isa<...>)
    return true;
  return false;
}

File: tools/onni/CountOperatorsPass.cpp [hunk 4]
```diff
@@ -27,6 +30,9 @@ Pass::ReturnType CountOperatorsPass::runOnModule(Module &pModule)
   uint64_t total = 0;
 
   for (ComputeOperator &cm : *pModule.getRootComputeGraph()) {
+    if (dyn_cast<InputOperator>(&cm)) continue;
```
Comment (cycheng) [inline tools/onni/CountOperatorsPass.cpp]: Replace with a function? IsIOWOperator() or SkipInCountStatistics()?

**Comment by cycheng** [comment]
LTGM! (But please keep fixing TODO list ;P)

File: tools/onni/InterpreterPass.cpp (hunk 2 above)
Comment (a127a127) [reply tools/onni/InterpreterPass.cpp]: Good advice!

File: tools/onni/InterpreterPass.cpp (hunk 1 above)
Comment (a127a127) [reply tools/onni/InterpreterPass.cpp]: ok~

File: tools/onni/InterpreterPass.cpp (hunk 3 above)
Comment (a127a127) [reply tools/onni/InterpreterPass.cpp]: ok~

File: tools/onni/CountOperatorsPass.cpp (hunk 4 above)
Comment (a127a127) [reply tools/onni/CountOperatorsPass.cpp]: Ok, that's Great! Thanks a lot!

**Pull request merged**